File: 5. godina/PARPRO/DZ 2/Rene/connect_4.py
from mpi4py import MPI
import numpy as np
import logging
import time

# naredba: mpiexec -n 8 -affinity -cores 1 python connect_4.py

ROWS = 6
COLUMNS = 7
EMPTY = 0
PLAYER = 1
AI = 2
DUBINA_ZADATAKA = 5
DUBINA_PRETRAZIVANJA = 6
#Mora vrijediti DUBINA_PRETRAZIVANJA >= DUBINA_ZADATAKA

# Postavljanje logiranja
logging.basicConfig(filename=f'connect_4_rank_{MPI.COMM_WORLD.Get_rank()}.log',
                    level=logging.DEBUG,
                    format='%(asctime)s - %(levelname)s - %(message)s',
                    filemode='w')

def odigraj_potez(ploca, column, player):
    obrnuti_poredak_redaka = reversed(range(ROWS))
    for row in obrnuti_poredak_redaka:
        if ploca[row][column] == EMPTY:
            ploca[row][column] = player
            return row, column
    return None, None

def evaluate_board_job(board):
    def is_player_winner(player):
        for row in range(3, ROWS):
            for column in range(COLUMNS - 3):
                if board[row][column] == player and board[row - 1][column + 1] == player and board[row - 2][column + 2] == player and board[row - 3][column + 3] == player:
                    return True

        for row in range(ROWS - 3):
            for column in range(COLUMNS):
                if board[row][column] == player and board[row + 1][column] == player and board[row + 2][column] == player and board[row + 3][column] == player:
                    return True

        for row in range(ROWS - 3):
            for column in range(COLUMNS - 3):
                if board[row][column] == player and board[row + 1][column + 1] == player and board[row + 2][column + 2] == player and board[row + 3][column + 3] == player:
                    return True
            
        for row in range(ROWS):
            for column in range(COLUMNS - 3):
                if board[row][column] == player and board[row][column + 1] == player and board[row][column + 2] == player and board[row][column + 3] == player:
                    return True

        return False
    
    if is_player_winner(PLAYER):
        #logging.debug(f"Player win is detected {board}")
        return -1
    elif is_player_winner(AI):
        #logging.debug(f"AI win is detected {board}")
        return 1
    else:
        return 0

def print_board(ploca):
    #logging.info(f"Trenutno stanje ploce:\n{ploca}")
    print(ploca)
    print()

def generiraj_poslove(board, broj_razina, current_path=[]): #rekurzivno generiranje svih poslova
    if not broj_razina:
        return [tuple(current_path)]
    
    jobs = []
    for column in range(COLUMNS):
        next_board = board.copy()
        player = AI if len(current_path) % 2 == 0 else PLAYER
        row, column = odigraj_potez(next_board, column, player)
        if row is None and column is None:
            continue
        jobs += generiraj_poslove(next_board, broj_razina - 1, current_path + [column])
    return jobs

def calculate_scores(current_result, broj_razina, is_ai_turn):
    if broj_razina == 0:
        return current_result

    parent_results = {}
    for path, score in current_result.items():
        parent_path = path[:-1]
        if parent_path not in parent_results:
            parent_results[parent_path] = []
        parent_results[parent_path].append(score)

    #generiranje scorea za cvorova po pravilima
    for parent_path, child_scores in parent_results.items():
        if is_ai_turn:
            if -1 in child_scores:
                #logging.debug(f"AI turn: Node {parent_path} has a child with -1")
                parent_results[parent_path] = -1
            else:
                avg_children_score = sum(child_scores) / len(child_scores)
                #logging.debug(f"AI turn: Node {parent_path} with child_scores {child_scores} average score is {avg_children_score}")
                parent_results[parent_path] = avg_children_score
        else:
            if 1 in child_scores:
                #logging.debug(f"Player turn: Node {parent_path} has a child with 1")
                parent_results[parent_path] = 1
            elif all(score == -1 for score in child_scores):
                #logging.debug(f"Player turn: All children of node {parent_path} have score -1")
                parent_results[parent_path] = -1
            else:
                #logging.debug(f"Player turn: child_scores {child_scores}")
                avg_children_score = sum(child_scores) / len(child_scores)
                #logging.debug(f"Player turn: Node {parent_path} average score is {avg_children_score}")
                parent_results[parent_path] = avg_children_score

    return calculate_scores(parent_results, broj_razina - 1, not is_ai_turn)

def worker():
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    #logging.info(f"Worker {rank} starting")
    status = MPI.Status()
    comm.send(None, dest=0, tag=0) # inicijalna prijava

    while True:
        message = comm.recv(source=0, tag=MPI.ANY_TAG, status=status)
        #logging.debug(f"Worker {rank} dobio posao {message}")
        if status.Get_tag() == 3:
            #logging.info(f"Worker {rank} primio poruku da nema vise poslova, proces zavrsava")
            break

        board, pathToTaskNode = message
        #logging.debug(f"Worker {rank} evaluating path {pathToTaskNode}")

        trenutni_player = AI  #AI je uvijek započinje
        for i in range(len(pathToTaskNode)):
            odigraj_potez(board, pathToTaskNode[i], trenutni_player)
            trenutni_player = PLAYER if trenutni_player == AI else AI  # Promjena igrača

        worker_results = {}
        worker_jobs = generiraj_poslove(board, DUBINA_PRETRAZIVANJA - DUBINA_ZADATAKA)
        #logging.info(f"Worker {rank} generated jobs: {worker_jobs}")

        if not worker_jobs: # nema vise mogucih slijednjih poslova ili DUBINA_PRETRAZIVANJA == DUBINA_ZADATAKA
            result = evaluate_board_job(board)
            comm.send((pathToTaskNode, result), dest=0, tag=1)
            #logging.debug(f"Worker {rank} salje masteru vrijednost za pathToTaskNode {pathToTaskNode} {result}")
            continue

        while worker_jobs:
            worker_path = worker_jobs.pop(0)
            #logging.debug(f"Worker {rank} uzima worker_job {worker_path}")
            temp_board = board.copy()
            trenutni_worker_player = trenutni_player
            for i in range(len(worker_path)):
                odigraj_potez(temp_board, worker_path[i], trenutni_worker_player)
                trenutni_worker_player = PLAYER if trenutni_worker_player == AI else AI  # Promjena igrača

            result = evaluate_board_job(temp_board)
            worker_results[pathToTaskNode + worker_path] = result
            #logging.debug(f"Worker {rank} completed worker_job {worker_path} with result {result}")
        
        is_ai_turn = (DUBINA_PRETRAZIVANJA % 2 == 0)  # ako to vrijedi AI je na potezu u cvorovima neposredno iznad listova
        
        final_results = calculate_scores(worker_results, DUBINA_PRETRAZIVANJA - DUBINA_ZADATAKA, is_ai_turn) # vraca vrijednost cvora pathToTaskNode
        
        score_to_send = None
        path_to_send = None

        for path, score in final_results.items(): # samo jedan entry ce bit i to za pathToTaskNode
            score_to_send = score
            path_to_send = path

        comm.send((path_to_send, score_to_send), dest=0, tag=1)
        #logging.debug(f"Worker {rank} salje masteru worker_job i rezultat {path_to_send} {score_to_send}")

def master():
    comm = MPI.COMM_WORLD
    size = comm.Get_size()
    rank = comm.Get_rank()
    #logging.info(f"Master {rank} starting with {size} processes")
    board = np.zeros((ROWS, COLUMNS), dtype=int)
    game_over = False
    workers_needing_job = []

    while not game_over:
        print_board(board)
        column = int(input(f"Select a column (0-6): "))
        odigraj_potez(board, column, PLAYER)

        start_time = time.perf_counter()

        if evaluate_board_job(board) == -1:
            print_board(board)
            print("Player wins")
            game_over = True
            break

        # Generiranje poslova za proizvoljnu dubinu
        jobs = generiraj_poslove(board, DUBINA_ZADATAKA)
        num_of_jobs = len(jobs)
        #logging.info(f"Master {rank} generated {num_of_jobs} jobs: {jobs}")
        results = {}

        if num_of_jobs == 0:
            #logging.info("Draw")
            game_over = True
            break

        status = MPI.Status()

        while len(results) < num_of_jobs:
            # Provjera dolaznih poruka
            while jobs and workers_needing_job:
                job = jobs.pop(0)
                free_worker = workers_needing_job.pop(0)
                #logging.debug(f"Sending job {job} to worker {free_worker}")
                comm.send((board.copy(), job), dest=free_worker, tag=1)
                
            if comm.Iprobe(source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG, status=status):
                message = comm.recv(source=MPI.ANY_SOURCE, tag=MPI.ANY_TAG, status=status)
                worker_rank = status.Get_source()
                #logging.debug(f"Received message from worker {worker_rank}")

                if status.Get_tag() == 0:  # Prijava radnika
                    if jobs:
                        job = jobs.pop(0)
                        comm.send((board.copy(), job), dest=worker_rank, tag=1)
                        #logging.debug(f"Master sent to worker {worker_rank} a job {job}")
                    else:
                        workers_needing_job.append(worker_rank)
                else:
                    #logging.debug(f"Worker {worker_rank} returned result: {message}")>
                    path, result = message
                    results[path] = result
                    if jobs:
                        job = jobs.pop(0)
                        comm.send((board.copy(), job), dest=worker_rank, tag=1)
                        #logging.debug(f"Master sent to worker {worker_rank} a job {job}")
                    else:
                        workers_needing_job.append(worker_rank)

            elif jobs: # master je implementiran da radi isto kao i workeri
                job = jobs.pop(0)
                #logging.debug(f"Master evaluating path {job}")
                temp_board = board.copy()
                pathToTaskNode = tuple(job)

                trenutni_player = AI  #AI je uvijek započinje
                for i in range(len(pathToTaskNode)):
                    odigraj_potez(temp_board, pathToTaskNode[i], trenutni_player)
                    trenutni_player = PLAYER if trenutni_player == AI else AI  # Promjena igrača

                master_as_worker_results = {}
                master_as_worker_jobs = generiraj_poslove(board, DUBINA_PRETRAZIVANJA - DUBINA_ZADATAKA)
                #logging.info(f"Master {rank} generated jobs: {master_as_worker_jobs}")

                if not master_as_worker_jobs: # nema vise mogucih slijednjih poslova ili DUBINA_PRETRAZIVANJA == DUBINA_ZADATAKA
                    result = evaluate_board_job(board)
                    results[pathToTaskNode] = result
                    #logging.debug(f"Master {rank} je obavio posao pathToTaskNode {pathToTaskNode} {result}")
                    continue

                while master_as_worker_jobs:
                    master_as_worker_path = master_as_worker_jobs.pop(0)
                    #logging.debug(f"Master {rank} uzima worker_job {master_as_worker_path}")
                    temp_temp_board = temp_board.copy()
                    trenutni_worker_player = trenutni_player
                    for i in range(len(master_as_worker_path)):
                        odigraj_potez(temp_temp_board, master_as_worker_path[i], trenutni_worker_player)
                        trenutni_worker_player = PLAYER if trenutni_worker_player == AI else AI  # Promjena igrača

                    result = evaluate_board_job(temp_temp_board)
                    master_as_worker_results[pathToTaskNode + master_as_worker_path] = result
                    #logging.debug(f"Master {rank} completed worker_job {master_as_worker_path} with result {result}")
                
                is_ai_turn = (DUBINA_PRETRAZIVANJA % 2 == 0)  # ako to vrijedi AI je na potezu u cvorovima neposredno iznad listova

                final_results = calculate_scores(master_as_worker_results, DUBINA_PRETRAZIVANJA - DUBINA_ZADATAKA, is_ai_turn) # vraca vrijednost cvora pathToTaskNode

                for pathToTaskNode, result in final_results.items(): # samo jedan entry ce bit i to za pathToTaskNode
                    results[pathToTaskNode] = result
                    #logging.debug(f"Master {rank} obavio posao {pathToTaskNode} {result}")

        is_ai_turn = (DUBINA_ZADATAKA % 2 == 0)  # ako to vrijedi AI je na potezu u cvorovima neposredno iznad listova
        #logging.debug(f"Initial results: {results}")
        final_results = calculate_scores(results, DUBINA_ZADATAKA - 1, is_ai_turn) # vrijednosti cvorova na prvoj razini ispod korijena
        #logging.debug(f"Final results: {final_results}")

        best_score = -2
        best_column = None

        for column, score in final_results.items():
            if score > best_score:
                best_score = score
                best_column = column[0]

        if best_column is not None:
            row, column = odigraj_potez(board, best_column, AI)
            if evaluate_board_job(board) == 1:
                print_board(board)
                print("AI wins")
                game_over = True
                break
        
        end_time = time.perf_counter()
        duration = end_time - start_time
        print(f"Trajanje izvrsavanja: {duration:.4f} sekundi")

    # Posalji svim workerima poruku da nema vise poslova
    for i in range(1, size):
        comm.send(None, dest=i, tag=3)

if __name__ == "__main__":
    comm = MPI.COMM_WORLD
    size = comm.Get_size()
    rank = comm.Get_rank()
    
    if rank == 0:
        master()
    else:
        worker()

File: 5. godina/PARPRO/DZ 2/connect4.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ROWS 6
#define COLUMNS 7
#define EMPTY 0
#define PLAYER 1
#define AI 2
#define DUBINA_ZADATAKA 5
#define DUBINA_PRETRAZIVANJA 6

void odigraj_potez(int ploca[ROWS][COLUMNS], int column, int player, int *row_result) {
    for (int row = ROWS - 1; row >= 0; row--) {
        if (ploca[row][column] == EMPTY) {
            ploca[row][column] = player;
            *row_result = row;
            return;
        }
    }
    *row_result = -1; // Indicating invalid move
}

int evaluate_board_job(int board[ROWS][COLUMNS]) {
    int is_player_winner(int player) {
        for (int row = 3; row < ROWS; row++) {
            for (int column = 0; column < COLUMNS - 3; column++) {
                if (board[row][column] == player && board[row - 1][column + 1] == player && 
                    board[row - 2][column + 2] == player && board[row - 3][column + 3] == player) {
                    return 1;
                }
            }
        }

        for (int row = 0; row < ROWS - 3; row++) {
            for (int column = 0; column < COLUMNS; column++) {
                if (board[row][column] == player && board[row + 1][column] == player && 
                    board[row + 2][column] == player && board[row + 3][column] == player) {
                    return 1;
                }
            }
        }

        for (int row = 0; row < ROWS - 3; row++) {
            for (int column = 0; column < COLUMNS - 3; column++) {
                if (board[row][column] == player && board[row + 1][column + 1] == player && 
                    board[row + 2][column + 2] == player && board[row + 3][column + 3] == player) {
                    return 1;
                }
            }
        }

        for (int row = 0; row < ROWS; row++) {
            for (int column = 0; column < COLUMNS - 3; column++) {
                if (board[row][column] == player && board[row][column + 1] == player && 
                    board[row][column + 2] == player && board[row][column + 3] == player) {
                    return 1;
                }
            }
        }

        return 0;
    }

    if (is_player_winner(PLAYER)) {
        return -1;
    } else if (is_player_winner(AI)) {
        return 1;
    } else {
        return 0;
    }
}

void print_board(int ploca[ROWS][COLUMNS]) {
    for (int row = 0; row < ROWS; row++) {
        for (int col = 0; col < COLUMNS; col++) {
            printf("%d ", ploca[row][col]);
        }
        printf("\n");
    }
    printf("\n");
    fflush(stdout);
}

void generiraj_poslove(int board[ROWS][COLUMNS], int broj_razina, int current_path[], int path_len, int jobs[][DUBINA_ZADATAKA], int *job_count) {
    if (broj_razina == 0) {
        memcpy(jobs[*job_count], current_path, path_len * sizeof(int));
        (*job_count)++;
        return;
    }

    for (int column = 0; column < COLUMNS; column++) {
        int next_board[ROWS][COLUMNS];
        memcpy(next_board, board, ROWS * COLUMNS * sizeof(int));
        int player = (path_len % 2 == 0) ? AI : PLAYER;
        int row_result;
        odigraj_potez(next_board, column, player, &row_result);
        if (row_result != -1) {
            current_path[path_len] = column;
            generiraj_poslove(next_board, broj_razina - 1, current_path, path_len + 1, jobs, job_count);
        }
    }
}

void calculate_scores(int current_result[][DUBINA_ZADATAKA], int scores[], int result_len, int broj_razina, int is_ai_turn) {
    if (broj_razina == 0) {
        return;
    }

    int parent_results[result_len][DUBINA_ZADATAKA];
    int parent_scores[result_len];
    int parent_len = 0;

    for (int i = 0; i < result_len; i++) {
        int parent_path_len = DUBINA_ZADATAKA - broj_razina + 1;
        memcpy(parent_results[parent_len], current_result[i], parent_path_len * sizeof(int));
        parent_scores[parent_len] = scores[i];
        parent_len++;
    }

    for (int i = 0; i < parent_len; i++) {
        if (is_ai_turn) {
            if (parent_scores[i] == -1) {
                scores[i] = -1;
            } else {
                scores[i] = 0; // Calculate average score
            }
        } else {
            if (parent_scores[i] == 1) {
                scores[i] = 1;
            } else if (parent_scores[i] == -1) {
                scores[i] = -1;
            } else {
                scores[i] = 0; // Calculate average score
            }
        }
    }

    calculate_scores(parent_results, scores, parent_len, broj_razina - 1, !is_ai_turn);
}

void worker() {
    MPI_Status status;
    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    printf("Worker %d starting\n", rank);
    fflush(stdout);
    MPI_Send(NULL, 0, MPI_INT, 0, 0, MPI_COMM_WORLD); // initial job request

    while (1) {
        int message[ROWS * COLUMNS + DUBINA_ZADATAKA];
        MPI_Recv(message, ROWS * COLUMNS + DUBINA_ZADATAKA, MPI_INT, 0, MPI_ANY_TAG, MPI_COMM_WORLD, &status);

        if (status.MPI_TAG == 3) {
            printf("Worker %d terminating\n", rank);
            fflush(stdout);
            break;
        }

        int board[ROWS][COLUMNS];
        memcpy(board, message, ROWS * COLUMNS * sizeof(int));

        int pathToTaskNode[DUBINA_ZADATAKA];
        memcpy(pathToTaskNode, &message[ROWS * COLUMNS], DUBINA_ZADATAKA * sizeof(int));

        int current_player = AI;
        for (int i = 0; i < DUBINA_ZADATAKA; i++) {
            if (pathToTaskNode[i] != -1) {
                odigraj_potez(board, pathToTaskNode[i], current_player, &current_player);
                current_player = (current_player == AI) ? PLAYER : AI;
            }
        }

        int worker_jobs[100][DUBINA_ZADATAKA];
        int worker_job_count = 0;
        int current_path[DUBINA_ZADATAKA] = {0};

        generiraj_poslove(board, DUBINA_PRETRAZIVANJA - DUBINA_ZADATAKA, current_path, 0, worker_jobs, &worker_job_count);

        if (worker_job_count == 0) {
            int result = evaluate_board_job(board);
            int result_message[DUBINA_ZADATAKA + 1];
            memcpy(result_message, pathToTaskNode, DUBINA_ZADATAKA * sizeof(int));
            result_message[DUBINA_ZADATAKA] = result;
            MPI_Send(result_message, DUBINA_ZADATAKA + 1, MPI_INT, 0, 1, MPI_COMM_WORLD);
        } else {
            int worker_results[100][DUBINA_ZADATAKA];
            int worker_scores[100];
            int worker_result_len = 0;

            for (int i = 0; i < worker_job_count; i++) {
                int temp_board[ROWS][COLUMNS];
                memcpy(temp_board, board, ROWS * COLUMNS * sizeof(int));

                int temp_path[DUBINA_ZADATAKA];
                memcpy(temp_path, worker_jobs[i], DUBINA_ZADATAKA * sizeof(int));

                int temp_current_player = current_player;
                for (int j = 0; j < DUBINA_ZADATAKA; j++) {
                    if (temp_path[j] != -1) {
                        odigraj_potez(temp_board, temp_path[j], temp_current_player, &temp_current_player);
                        temp_current_player = (temp_current_player == AI) ? PLAYER : AI;
                    }
                }

                worker_scores[worker_result_len] = evaluate_board_job(temp_board);
                memcpy(worker_results[worker_result_len], worker_jobs[i], DUBINA_ZADATAKA * sizeof(int));
                worker_result_len++;
            }

            int is_ai_turn = (DUBINA_PRETRAZIVANJA % 2 == 0);

            calculate_scores(worker_results, worker_scores, worker_result_len, DUBINA_PRETRAZIVANJA - DUBINA_ZADATAKA, is_ai_turn);

            int score_to_send = worker_scores[0];
            int path_to_send[DUBINA_ZADATAKA];
            memcpy(path_to_send, worker_results[0], DUBINA_ZADATAKA * sizeof(int));

            int result_message[DUBINA_ZADATAKA + 1];
            memcpy(result_message, pathToTaskNode, DUBINA_ZADATAKA * sizeof(int));
            result_message[DUBINA_ZADATAKA] = score_to_send;
            MPI_Send(result_message, DUBINA_ZADATAKA + 1, MPI_INT, 0, 1, MPI_COMM_WORLD);
        }
    }
}

void master() {
    int board[ROWS][COLUMNS] = {0};
    int game_over = 0;
    int size;
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    printf("Master starting with %d processes\n", size);
    fflush(stdout);

    while (!game_over) {
        print_board(board);
        int column;
        printf("Select a column (0-6): ");
        fflush(stdout);
        scanf("%d", &column);

        int row_result;
        odigraj_potez(board, column, PLAYER, &row_result);

        if (evaluate_board_job(board) == -1) {
            print_board(board);
            printf("Player wins\n");
            fflush(stdout);
            game_over = 1;
            break;
        }

        int jobs[100][DUBINA_ZADATAKA];
        int job_count = 0;
        int current_path[DUBINA_ZADATAKA] = {0};
        generiraj_poslove(board, DUBINA_ZADATAKA, current_path, 0, jobs, &job_count);

        int results[100];
        int result_count = 0;

        MPI_Status status;
        while (result_count < job_count) {
            for (int i = 0; i < job_count; i++) {
                int message[ROWS * COLUMNS + DUBINA_ZADATAKA];
                memcpy(message, board, ROWS * COLUMNS * sizeof(int));
                memcpy(&message[ROWS * COLUMNS], jobs[i], DUBINA_ZADATAKA * sizeof(int));
                MPI_Send(message, ROWS * COLUMNS + DUBINA_ZADATAKA, MPI_INT, i % (size - 1) + 1, 1, MPI_COMM_WORLD);
            }

            for (int i = 0; i < job_count; i++) {
                int result_message[DUBINA_ZADATAKA + 1];
                MPI_Recv(result_message, DUBINA_ZADATAKA + 1, MPI_INT, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &status);
                int path[DUBINA_ZADATAKA];
                memcpy(path, result_message, DUBINA_ZADATAKA * sizeof(int));
                results[result_count] = result_message[DUBINA_ZADATAKA];
                result_count++;
            }
        }

        int is_ai_turn = (DUBINA_ZADATAKA % 2 == 0);
        int final_results[100];
        calculate_scores(jobs, results, result_count, DUBINA_ZADATAKA - 1, is_ai_turn);

        int best_score = -2;
        int best_column = -1;
        for (int i = 0; i < result_count; i++) {
            if (results[i] > best_score) {
                best_score = results[i];
                best_column = jobs[i][0];
            }
        }

        if (best_column != -1) {
            odigraj_potez(board, best_column, AI, &row_result);
            if (evaluate_board_job(board) == 1) {
                print_board(board);
                printf("AI wins\n");
                fflush(stdout);
                game_over = 1;
                break;
            }
        }

        printf("Round completed\n");
        fflush(stdout);
    }

    for (int i = 1; i < size; i++) {
        MPI_Send(NULL, 0, MPI_INT, i, 3, MPI_COMM_WORLD);
    }
}

int main(int argc, char *argv[]) {
    MPI_Init(&argc, &argv);

    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    if (rank == 0) {
        master();
    } else {
        worker();
    }

    MPI_Finalize();
    return 0;
}
